src/state.rs

```diff
@@ -0,0 +1,211 @@
+        self.memory_stack_state.chain_id()
+    }
+
+    fn exists(&self, address: H160) -> bool {
+        self.memory_stack_state.exists(address)
+    }
+
+    fn basic(&self, address: H160) -> Basic {
+        self.memory_stack_state.basic(address)
+    }
+
+    fn code(&self, address: H160) -> Vec<u8> {
+        self.memory_stack_state.code(address)
+    }
+
+    fn storage(&self, address: H160, index: H256) -> H256 {
+        self.memory_stack_state.storage(address, index)
+    }
+
+    fn original_storage(&self, address: H160, index: H256) -> Option<H256> {
+        self.memory_stack_state.original_storage(address, index)
+    }
+}
+
+impl<'backend, 'config> StackState<'config> for AuroraStackState<'backend, 'config> {
+    fn metadata(&self) -> &StackSubstateMetadata<'config> {
+        self.memory_stack_state.metadata()
+    }
+
+    fn metadata_mut(&mut self) -> &mut StackSubstateMetadata<'config> {
+        self.memory_stack_state.metadata_mut()
+    }
+
+    fn enter(&mut self, gas_limit: u64, is_static: bool) {
+        // TODO: What should we do with respect to gas limit
+        self.promises.enter();
+
+        self.memory_stack_state.enter(gas_limit, is_static);
+    }
+
+    fn exit_commit(&mut self) -> Result<(), ExitError> {
+        self.promises.commit();
+
+        self.memory_stack_state.exit_commit()
+    }
+
+    fn exit_revert(&mut self) -> Result<(), ExitError> {
+        self.promises.discard();
+
+        self.memory_stack_state.exit_revert()
+    }
+
+    fn exit_discard(&mut self) -> Result<(), ExitError> {
+        self.promises.discard();
+
+        self.memory_stack_state.exit_discard()
+    }
+
+    fn is_empty(&self, address: H160) -> bool {
+        self.memory_stack_state.is_empty(address)
+    }
+
+    fn deleted(&self, address: H160) -> bool {
+        self.memory_stack_state.deleted(address)
+    }
+
+    fn inc_nonce(&mut self, address: H160) {
+        self.memory_stack_state.inc_nonce(address)
+    }
+
+    fn set_storage(&mut self, address: H160, key: H256, value: H256) {
+        self.memory_stack_state.set_storage(address, key, value)
+    }
+
+    fn reset_storage(&mut self, address: H160) {
+        self.memory_stack_state.reset_storage(address)
+    }
+
+    fn log(&mut self, address: H160, topics: Vec<H256>, data: Vec<u8>) {
+        self.memory_stack_state.log(address, topics, data)
+    }
+
+    fn set_deleted(&mut self, address: H160) {
+        self.memory_stack_state.set_deleted(address)
+    }
+
+    fn set_code(&mut self, address: H160, code: Vec<u8>) {
+        self.memory_stack_state.set_code(address, code)
+    }
+
+    fn transfer(&mut self, transfer: Transfer) -> Result<(), ExitError> {
+        self.memory_stack_state.transfer(transfer)
+    }
+
+    fn reset_balance(&mut self, address: H160) {
+        self.memory_stack_state.reset_balance(address)
+    }
+
+    fn touch(&mut self, address: H160) {
+        self.memory_stack_state.touch(address)
+    }
+}
+
+struct Stack<T> {
```

Comment:
I'd love to see more tests with this around `enter`, `commit`, `discard` (especially with multiple nesting of `enter`).